### PR TITLE
feat: Klarna headless RN (Android & iOS wrapper, TS layer, example screen)

### PR DESCRIPTION
**Pull request by FlaviuExtPrimer**
## Description

https://primer-ext.atlassian.net/browse/TWS-94
https://primer-ext.atlassian.net/browse/TWS-95
https://primer-ext.atlassian.net/browse/TWS-165

- Implement wrapper for Android's KlarnaPaymentComponent & PrimerHeadlessUniversalCheckoutKlarnaManager **(iOS wrapper will be done separately)**
- Implement TypeScript public API layer
- Implement example screen for Klarna headless
- Implement session intent switcher on checkout screen
- Updated Gradle, and kotlinx.serialization

👉 Affected payment methods:
- [ ] Card
- [x] Klarna
- [ ] PayPal
- [ ] GoCardless
- [ ] Google Pay

### Public API Changes
- Created public API for Klarna, see KlarnaManager.ts, and index.tsx changes

## How has this been tested?
- manually

## Screenshots:

https://github.com/primer-io/primer-sdk-react-native/assets/149587774/362db013-1844-4801-8264-cfa5e9b97149

## Checklist:

- [ ] Documentation has been updated or required updates are tracked (link 
relevant tickets)
- [x] Tested flows where relevant
- [ ] Tests added/updated 
- [x] My code follows the code style of this project

### CODE REVIEW

**Comment by FlaviuExtPrimer** [comment]
Will not compile because Klarna SDK for Android hasn't been merged.

TODO:
- [ ] : update Klarna SDK version once Klarna SDK for Android is merged.

**Review by HarryPrimer — APPROVED** [review]
LGTM

File: packages/sdk/android/src/main/java/com/primerioreactnative/components/manager/klarna/PrimerRNHeadlessUniversalCheckoutKlarnaComponent.kt [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+        }
+    }
+
+    private fun JSONObject?.toWritableMap(): WritableMap =
+        this?.let { convertJsonToMap(this) } ?: Arguments.createMap()
+
+    private fun JSONArray?.toWritableArray(): WritableArray =
+        this?.let { convertJsonToArray(this) } ?: Arguments.createArray()
+
+    private fun sendEvent(name: String, data: JSONObject?) {
```
Comment (semirp) [inline packages/sdk/android/src/main/java/com/primerioreactnative/components/manager/klarna/PrimerRNHeadlessUniversalCheckoutKlarnaComponent.kt:311]: [nit]: can we use extension functions from `JsonConverter` class in utils package?

File: packages/sdk/android/src/main/java/com/primerioreactnative/components/manager/klarna/PrimerRNHeadlessUniversalCheckoutKlarnaComponent.kt [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+        if (klarnaComponent == null) {
+            val exception = ErrorTypeRN.NativeBridgeFailed errorTo UNINITIALIZED_ERROR
+            promise.reject(exception.errorId, exception.description)
+        } else {
+            klarnaComponent?.updateCollectedData(KlarnaPaymentCollectableData.FinalizePayment)
+            promise.resolve(null)
+        }
+    }
+
+    private fun JSONObject?.toWritableMap(): WritableMap =
```
Comment (semirp) [inline packages/sdk/android/src/main/java/com/primerioreactnative/components/manager/klarna/PrimerRNHeadlessUniversalCheckoutKlarnaComponent.kt:305]: [nit]: can we use extension functions from `JsonConverter` class in utils package?

**Review by semirp — APPROVED** [review]
@FlaviuExtPrimer Few nits, but Android LGTM!

Would be nice to add ktlint to android module, as we have a lot of newly added classes that are added and not ended with newline char.

**Comment by FlaviuExtPrimer** [comment]
> @FlaviuExtPrimer Few nits, but Android LGTM!
> 
> Would be nice to add ktlint to android module, as we have a lot of newly added classes that are added and not ended with newline char.

Would this be done at CI level too?

File: packages/sdk/android/src/main/java/com/primerioreactnative/components/manager/klarna/PrimerRNHeadlessUniversalCheckoutKlarnaComponent.kt (hunk 1 above)
Comment (FlaviuExtPrimer) [reply packages/sdk/android/src/main/java/com/primerioreactnative/components/manager/klarna/PrimerRNHeadlessUniversalCheckoutKlarnaComponent.kt:305]: Updated

File: packages/sdk/android/src/main/java/com/primerioreactnative/components/manager/klarna/PrimerRNHeadlessUniversalCheckoutKlarnaComponent.kt (hunk 2 above)
Comment (FlaviuExtPrimer) [reply packages/sdk/android/src/main/java/com/primerioreactnative/components/manager/klarna/PrimerRNHeadlessUniversalCheckoutKlarnaComponent.kt]: Updated

**Comment by FlaviuExtPrimer** [comment]
> Will not compile because Klarna SDK for Android hasn't been merged.
> 
> TODO:
> 
>     * [x]  : update Klarna SDK version once primer.android 2.26 is merged in master.

I ended up upgrading the main SDK and Klarna-android on this branch.

File: packages/example/src/screens/HeadlessCheckoutVaultScreen.tsx [hunk 3]
```diff
@@ -278,13 +284,23 @@ export default HeadlessCheckoutVaultScreen = (props: any) => {
...
+  const getVaultedPaymentData = (item: any) => {
+      const last4Digits = item.paymentInstrumentData.last4Digits
+      if (last4Digits !== undefined) {
+        return '••••' + last4Digits
+      }
+      const email = item.paymentInstrumentData?.sessionData?.billingAddress?.email ?? ""
+      const paymentMethodType = item?.paymentMethodType ?? ""
+      return paymentMethodType + ": " + email;
+  }
+
```
Comment (semirp) [inline packages/example/src/screens/HeadlessCheckoutVaultScreen.tsx:303]: [nit]: we could switch over the `paymentMethodType` here.

**Review by semirp — APPROVED** [review]

File: packages/example/src/screens/HeadlessCheckoutVaultScreen.tsx (hunk 3 above)
Comment (semirp) [reply packages/example/src/screens/HeadlessCheckoutVaultScreen.tsx:303]: @FlaviuExtPrimer we are trying to display relevant data for saved payment method.

Comment (FlaviuExtPrimer) [reply packages/example/src/screens/HeadlessCheckoutVaultScreen.tsx]: Changes pushed.

**Review by NQuinn27 — APPROVED** [review]

**Pull request merged**